### ISSUE TITLE
Support jax.random operations.

### ISSUE THREAD

**Issue by DAOvb** [enpasos/jax2onnx#11]
Greetings! Thank you for your effort! I really appreciate your work! What do you think about supporting jax.random operations? Some coversions like jax.random.multinomial -> onnx.multinomial and the same for normal? I think they could be useful is some cases (like Diffusion models, RL, etc). May be I can help you with that, but I don't know currently how to convert jax.random.PRNGKey -> onnx seed which is required in such operations.

**Comment by enpasos** [comment]
Welcome @DAOvb, great to have you here.

Supporting jax.random operations sounds like a valuable addition. Plugins could fit under jax2onnx/plugins/jax/random.

I have to admit I’m not very familiar with this topic myself … but my LLM seems to have an opinion: "the main challenge is converting JAX’s PRNGKey into an ONNX-compatible seed, since JAX uses a stateful two-element key, while ONNX expects a numeric seed. But this should be solvable with a consistent mapping."

Would be great if you want to explore this — happy to support.

**Comment by enpasos** [comment]
<p>@DAOvb: As you mentioned, getting JAX and ONNX Runtime in sync for PRNG and seed-based deterministic behavior is likely a challenge. If you'd like to give it a try, you're very welcome to do so.</p>
<p>Here are some thoughts (LLM-assisted, so take them with a grain of salt) on how to approach <strong>seed extraction and PRNG alignment between JAX and ONNX Runtime</strong>:</p>
<hr>
<h3>1. Understanding JAX’s PRNGKey</h3>
<ul>
<li>
<p>JAX uses a <strong>stateless</strong>, functional PRNG system based on a 2-element <code>uint32</code> array.</p>
</li>
<li>
<p>Each random op takes a PRNGKey and returns a new one, ensuring reproducibility without side effects.</p>
</li>
</ul>
<h3>2. ONNX Random Operators</h3>
<ul>
<li>
<p>ONNX has random ops like <code>RandomNormal</code>, <code>RandomUniform</code>, <code>Multinomial</code>, etc.</p>
</li>
<li>
<p>These accept a single <code>seed</code> attribute (float or int), not a key array.</p>
</li>
</ul>
<h3>3. Mapping JAX PRNGKey → ONNX Seed</h3>
<p>To deterministically extract a seed from a PRNGKey:</p>
<pre><code class="language-python">def prngkey_to_seed(key):
    return int(key[0]) &lt;&lt; 32 | int(key[1])
</code></pre>
<ul>
<li>
<p>This creates a consistent mapping from JAX PRNGKeys to ONNX seeds.</p>
</li>
<li>
<p>Use this when emitting ONNX random nodes.</p>
</li>
</ul>
<h3>4. Ensuring Reproducibility</h3>
<ul>
<li>
<p>When exporting the model, use a fixed PRNGKey.</p>
</li>
<li>
<p>Pass the corresponding extracted seed as the <code>seed</code> attribute in the ONNX node.</p>
</li>
</ul>
<h3>5. Limitations</h3>
<ul>
<li>
<p><strong>JAX’s PRNG supports splitting</strong> (<code>jax.random.split</code>), but ONNX does not.<br>
You’ll need to mimic this logic or document that reproducibility only applies to the initial seed.</p>
</li>
<li>
<p><strong>ONNX seeds are floats</strong>, which may lead to minor precision loss for large integers.<br>
Usually not an issue, but worth noting.</p>
</li>
</ul>
<h3>6. Example: Plugin for RandomNormal</h3>
<pre><code class="language-python">class RandomNormalPlugin(PrimitiveLeafPlugin):
    def to_onnx(self, converter, node_inputs, node_outputs, params):
        prng_key = node_inputs[0]
        shape = ...  # infer from context
        seed = prngkey_to_seed(prng_key)
        node = onnx.helper.make_node(
            "RandomNormal",
            inputs=[],
            outputs=[converter.get_name(node_outputs[0])],
            shape=shape,
            dtype=onnx.TensorProto.FLOAT,
            seed=float(seed),
        )
        converter.add_node(node)
</code></pre>
<h3>7. User Guidance</h3>
<ul>
<li>
<p>Document this:<br>
“To ensure reproducible random results across JAX and ONNX Runtime, pass a fixed PRNGKey to your JAX model and map it to a numeric ONNX seed using the method above.”</p>
</li>
</ul>
 <hr>

<p>Let me know if you'd like to dive in and bootstrap a plugin under jax2onnx/plugins/jax/random — happy to help get you started if you want to explore that direction..</p>
